incubator/hnc/internal/validators/object.go

```diff
@@ -244,16 +247,17 @@ func (o *Object) handleInherited(op admissionv1beta1.Operation, newSource, oldSo
 	return deny(metav1.StatusReasonInternalError, "unknown operation: "+string(op))
 }
 
-// isDeletingNS returns true if the namespace of the object is already being deleted, based on the
-// in-memory forest.
-func (o *Object) isDeletingNS(inst *unstructured.Unstructured) bool {
-	o.Forest.Lock()
-	defer o.Forest.Unlock()
-	ns := o.Forest.Get(inst.GetNamespace())
-	if ns == nil {
-		return false
+// isDeletingNS returns true if the namespace of the object is already being deleted.
+func (o *Object) isDeletingNS(ctx context.Context, inst *unstructured.Unstructured) bool {
+	ns := &corev1.Namespace{}
+	nnm := types.NamespacedName{Name: inst.GetNamespace()}
+	err := o.client.Get(ctx, nnm, ns)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return false
+		}
```

Comment:
Now that I think about it, IsNotFound will return true if the namespace was _previously_ deleted, not if it's _being_ deleted. But if the namespace is already gone, there won't be any objects in it so this function will never be called.

So I think that if there are any errors returned from the client, we should just return it directly to the caller without caring what it is.

The way to tell if the namespace is actually _being_ deleted is `if !ns.GetDeletionTimestamp().IsZero()`. We set the deletion timestamp when we _start_ deleting an object.